core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1410,10 +1410,12 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     val scheme = new URI(schemeCorrectedPath).getScheme
     if (!Array("http", "https", "ftp").contains(scheme)) {
       val fs = hadoopPath.getFileSystem(hadoopConfiguration)
-      if (!fs.exists(hadoopPath)) {
-        throw new FileNotFoundException(s"Added file $hadoopPath does not exist.")
+      val isDir = try {
+        fs.getFileStatus(hadoopPath).isDirectory
+      } catch {
+        case f: FileNotFoundException =>
+          throw new FileNotFoundException(s"Added file $hadoopPath does not exist.").initCause(f)
```

Comment:
BTW, if you look at how FileSystem.exists() is implemented' it's getFileStatus with the catch downgraded to returning false. Against S3, the status check is taking 1-3 HTTPS requests, so measurably slow